File: lib/Backend/repositories/MenuRepository.js
const db = require('../patterns/singletons/DatabaseManager');

class MenuRepository {
  async findAll({ categoryId, availability, search } = {}) {
    let sql = `SELECT m.*, c.category_name FROM menu_items m
               JOIN menu_categories c ON m.category_id = c.category_id WHERE 1=1`;
    const params = [];
    if (categoryId) { sql += ' AND m.category_id = ?'; params.push(categoryId); }
    if (availability) { sql += ' AND m.availability = ?'; params.push(availability); }
    if (search) { sql += ' AND m.food_name LIKE ?'; params.push(`%${search}%`); }
    sql += ' ORDER BY c.category_name, m.food_name';
    return db.query(sql, params);
  }

  async findById(id) {
    const rows = await db.query('SELECT * FROM menu_items WHERE menu_id = ?', [id]);
    return rows[0] || null;
  }

  async create(data) {
    const { categoryId, foodName, description, price, image } = data;
    const result = await db.query(
      `INSERT INTO menu_items (category_id, food_name, description, price, image) VALUES (?, ?, ?, ?, ?)`,
      [categoryId, foodName, description, price, image || null]
    );
    return this.findById(result.insertId);
  }

  async update(id, fields) {
    const keys = Object.keys(fields);
    if (!keys.length) return this.findById(id);
    const setClause = keys.map((k) => `${k} = ?`).join(', ');
    await db.query(`UPDATE menu_items SET ${setClause} WHERE menu_id = ?`, [...Object.values(fields), id]);
    return this.findById(id);
  }

  async delete(id) {
    await db.query('DELETE FROM menu_items WHERE menu_id = ?', [id]);
  }

  async setAvailability(id, availability) {
    await db.query('UPDATE menu_items SET availability = ? WHERE menu_id = ?', [availability, id]);
    return this.findById(id);
  }

  // Categories
  async findAllCategories() {
    return db.query('SELECT * FROM menu_categories ORDER BY category_name');
  }

  async createCategory(categoryName) {
    const result = await db.query('INSERT INTO menu_categories (category_name) VALUES (?)', [categoryName]);
    return { category_id: result.insertId, category_name: categoryName };
  }

  async deleteCategory(id) {
    await db.query('DELETE FROM menu_categories WHERE category_id = ?', [id]);
  }
}

module.exports = new MenuRepository();

File: lib/Frontend/repositories/menu_repository.dart
import 'package:rms/Frontend/models/menu.dart';
import 'package:rms/Frontend/services/api_services.dart';

class MenuRepository {
  final ApiService _api = ApiService();

  Future<List<MenuItem>> getAll({String? search, String? availability}) async {
    final data = await _api.get('/menu', query: {
      if (search != null) 'search': search,
      if (availability != null) 'availability': availability,
    });
    return (data as List).map((e) => MenuItem.fromJson(e)).toList();
  }

  Future<List<Map<String, dynamic>>> getCategories() async {
    final data = await _api.get('/menu/categories');
    return (data as List).cast<Map<String, dynamic>>();
  }

  Future<Map<String, dynamic>> createCategory(String categoryName) async {
    final data = await _api.post('/menu/categories', body: {
      'categoryName': categoryName,
    });
    return Map<String, dynamic>.from(data);
  }

  /// Uploads an image and returns its relative URL (e.g. "/uploads/xyz.jpg").
  /// Pass this straight into `create(...)`'s `image` argument.
  Future<String> uploadImage({required List<int> bytes, required String filename}) async {
    final data = await _api.uploadFile('/menu/upload-image', bytes: bytes, filename: filename);
    return data['imageUrl'] as String;
  }

  Future<MenuItem> create({
    required int categoryId,
    required String foodName,
    required String description,
    required double price,
    String? image,
  }) async {
    final data = await _api.post('/menu', body: {
      'categoryId': categoryId,
      'foodName': foodName,
      'description': description,
      'price': price,
      if (image != null) 'image': image,
    });
    return MenuItem.fromJson(data);
  }

  Future<void> setAvailability(int menuId, bool available) {
    return _api.patch('/menu/$menuId/availability', body: {
      'availability': available ? 'available' : 'unavailable',
    });
  }

  Future<void> delete(int menuId) => _api.delete('/menu/$menuId');
}